### ISSUE TITLE
[New Concept Docs]:  Class Inheritance

### ISSUE THREAD

**Issue by BethanyG** [exercism/python#2972]
This issue describes how to implement the `class-inheritance` concept docs.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [Contributing to Exercism](https://exercism.org/docs/building) | [Exercism and GitHub](https://exercism.org/docs/building/github) | - [Contributor Pull Request Guide](https://exercism.org/docs/building/github/contributors-pull-request-guide)
- [What are those Weird Task Tags about?](https://exercism.org/docs/building/product/tasks)
- [Building Language Tracks: An Overview](https://exercism.org/docs/building/tracks) 
- [What are Concepts?](https://exercism.org/docs/building/tracks/concepts)
- [Concept Specifications](https://exercism.org/docs/building/tracks/concepts) 
- [Exercism Formatting and Style Guide](https://exercism.org/docs/building/markdown/style-guide)
- [Exercism Markdown Specification](https://exercism.org/docs/building/markdown/markdown)
- [Reputation](https://exercism.org/docs/using/product/reputation)  

## Goal

This concept is meant to teach an understanding/use of `class-inheritance`  in Python. 

## Learning objectives

- more fully understand the uses (_and possible abuses_) of `classes`  in Python
- understanding and use of `single inheritance` in Python
- understanding and use of `multiple inheritance` in Python
  - the diamond inheritance problem
  - Pythons MRO (_method resolution order_)
  - pros and cos of `mixins` () to isolate `multiple inheritance` issues
- when and when not to use `__super__()` to call a parent `class` constructor

  

## Out of scope

- `class-composition` as an explicit topic (_it is ok to use limited composition as needed in example code_)
- `dataclasses`
- `decorators` outside of `@property`, `@staticmethod`, and `@classmethod` (_the student should already be exposed to these in the prerequisite exercise_)
- `generators`
- `coroutines`
- `descriptors` (_these will get their own exercise_)
- using a `class` as a decorator
- performance considerations

  
## Concepts

- `classes`
- `inheritance` in `OOP`
- `single inheritance`
- `multiple inheritance`
- `minxins`
- `__super__()`

## Prerequisites

These are the concepts/concept exercises the student needs to complete/understand before solving the concept exercise that these concept documents support.

- `basics`
- `booleans`
- `classes`
- `class-customization`
- `comparisons`
- `decorators`
- `dicts`
- `functions`
- `higher-order-functions`
- `iteration`
- `lists`
- `numbers`
- `sequences`
- `sets`
- `strings`
- `tuples`

  

## Resources to refer to

- [classes (Python tutorial)](https://docs.python.org/3/tutorial/classes.html)
- [Real Python: Object-Oriented Programming in Python 3](https://realpython.com/python3-object-oriented-programming/)
- [Real Python: Inheritance and Composition: A Python OOP Guide](https://realpython.com/inheritance-composition-python/)
- [Towards DataScience: Understand Inheritance in Python](https://towardsdatascience.com/understand-inheritance-in-python-74f8e3025f3c) -- (_this is not a good resource for hints or links, since it requires an account signup_)
  - [Liskov substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle)
- [Multiple Inheritance in Python](https://www.python-course.eu/python3_multiple_inheritance.php)
- [Multiple Inheritance is Hard (Ned Batchelder)](https://nedbatchelder.com/blog/201210/multiple_inheritance_is_hard.html)
- [Python.org: MRO in Python 2.3+](https://www.python.org/download/releases/2.3/mro/)
- [Dig into Python super() and MRO](https://gaopinghuang0.github.io/2018/12/29/dig-into-python-super-and-mro)
- [Multiple inheritance and mixin classes in Python](https://www.thedigitalcatonline.com/blog/2020/03/27/mixin-classes-in-python/)
- [Making Python classes more modular using mixins](https://www.residentmar.io/2019/07/07/python-mixins.html)
- [Write composable, reusable Python classes using Mixinsl)](https://dev.to/bikramjeetsingh/write-composable-reusable-python-classes-using-mixins-6lj)
- [Mixins in Python and Ruby Compared](https://andrewbrookins.com/technology/mixins-in-python-and-ruby-compared/)
- [Python Data Model - Python Docs](https://docs.python.org/3/reference/datamodel.html#the-standard-type-hierarchy)

* ### `links.json`

  For more information, see [concept links file](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-linksjson)

  -  The same resources listed in this issue can be used as a starting point for the [ `concepts/links.json`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-linksjson)  file, if it doesn't already exist.
  -  If there are particularly good/interesting information sources for this concept that extend or supplement the concept exercise material & the resources already listed -- please add them to the `links.json` document.

## Concept Description

Please see the following for more details on these files:  [concepts](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-linksjson)

* ### Concept `about.md`

  For more information, see [Concept `about.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-aboutmd)

  - This file provides information about this concept for a student who has completed the corresponding concept exercise.  It is intended as a reference for continued learning.  

* ### Concept `introduction.md`

  For more information, see [Concept `introduction.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-introductionmd)

  * This can also be a summary/paraphrase of the  document listed above, and will provide a brief introduction of the concept for a student who has **not yet** completed the concept exercise.   It should contain a good summation of the concept, but not go into lots of detail.

## Implementation Notes

- Example code  should **only use syntax & concepts introduced within these docs or one of the prerequisite concept exercises or documents.**
  Please **do not use** comprehensions, generator expressions, or other syntax not previously covered.  Please also follow [PEP8](https://www.python.org/dev/peps/pep-0008/) guidelines.
- Our markdown and JSON files are checked against [prettier](https://prettier.io/) .  We recommend [setting prettier up locally](https://prettier.io/docs/en/install.html) and running it prior to submitting your PR  to avoid any CI errors.

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue, or contact one of the maintainers on our Slack channel.

**Comment by BethanyG** [comment]
Linking PR #2960

**Comment by BethanyG** [comment]
@[girijakar](https://github.com/girijakar) - not able to directly assign you, but this is an issue for just the concept docs.  😄

**Comment by girijakar** [comment]
Ok thank you

**Comment by BethanyG** [comment]
Once you commented, I could assign.  So now you are assigned.  😉

**Comment by BethanyG** [comment]
Closing, as the PR for this has been optimistically merged.  Further changes and improvements will be tracked under an improvement issue.  Many thanks for your work on this, @girijakar!
